src/cloudflare/workers.ts

```diff
@@ -6,9 +6,92 @@
 //   wrapper module that simply re-exports the classes from the built-in module.
 
 import entrypoints from 'cloudflare-internal:workers';
+import innerEnv from 'cloudflare-internal:env';
 
 export const WorkerEntrypoint = entrypoints.WorkerEntrypoint;
 export const DurableObject = entrypoints.DurableObject;
 export const RpcStub = entrypoints.RpcStub;
 export const RpcTarget = entrypoints.RpcTarget;
 export const WorkflowEntrypoint = entrypoints.WorkflowEntrypoint;
+
+// We turn off these specific linting rules here intentionally as mixing
+// typescript and proxies can just be... awkward. The use of any in these
+// is intentional.
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+export function withEnv(newEnv: unknown, fn: () => unknown): unknown {
+  return innerEnv.withEnv(newEnv, fn);
+}
+
+// A proxy for the workers env/bindings. The proxy is io-context
+// aware in that it will only return values when there is an active
+// IoContext. Mutations to the env via this proxy propagate to the
+// underlying env and do not follow the async context.
+export const env: object = new Proxy(
+  {},
+  {
+    get(_: any, prop: string | symbol): unknown {
+      const inner = innerEnv.getCurrent();
+      if (inner) {
+        return Reflect.get(inner, prop);
+      }
+      return undefined;
+    },
+
+    set(_: any, prop: string | symbol, newValue: any): boolean {
+      const inner = innerEnv.getCurrent();
+      if (inner) {
+        return Reflect.set(inner, prop, newValue);
+      }
+      return true;
+    },
+
+    has(_: any, prop: string | symbol): boolean {
+      const inner = innerEnv.getCurrent();
+      if (inner) {
+        return Reflect.has(inner, prop);
+      }
+      return false;
+    },
+
+    ownKeys(_: any): any {
+      const inner = innerEnv.getCurrent();
+      if (inner) {
+        return Reflect.ownKeys(inner);
+      }
+      return [];
+    },
+
+    deleteProperty(_: any, prop: string | symbol): boolean {
+      const inner = innerEnv.getCurrent();
+      if (inner) {
+        return Reflect.deleteProperty(inner, prop);
+      }
+      return true;
```

Comment:
Should this return `false` here ?

Reply:
Wait... actually no. `true` is correct here. The reason is that this is effectively deleting a non-existent property. If, for instance, you have `const n = {}; delete n.abc`, that delete will return `true` because effectively the property *is* deleted.

Reply:
Ok, I was looking at a MDN [snippet](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy#a_complete_traps_list_example):

```ts
const docCookies = new Proxy(docCookies, {
// ...
  deleteProperty(target, key) {
    if (!(key in target)) {
      return false;
    }
    return target.removeItem(key);
  },
```

Reply:
There's some disagreement in the ecosystem about exactly what should be returned in which cases but I tend to err on duplicating what I see v8 doing which is returning `true` when deleting a non-existent property.

Reply:
It looks like returning `false` might throw in strict mode, which we do not want.

Maybe a simpler impl for all method could be:

```ts
// before
      const inner = innerEnv.getCurrent();
      if (inner) {
        return Reflect.deleteProperty(inner, prop);
      }
      return true;
      
// after
return Reflect.deleteProperty( innerEnv.getCurrent() ?? {}, prop);
```

Reply:
I think I prefer the more verbose version as I think it's a bit clearer / easier to read. It will also be a bit faster as it avoids the extraneous property lookup on the empty object.